AX: Add fast path for text marker ordering when one marker is the start of the root web area

hoffmanjoshua opened the pull request:
#### d49d59b1464d5196a8fafae70423b3e9ded4830b
<pre>
AX: Add fast path for text marker ordering when one marker is the start of the root web area
<a href="https://bugs.webkit.org/show_bug.cgi?id=289034">https://bugs.webkit.org/show_bug.cgi?id=289034</a>
<a href="https://rdar.apple.com/146075994">rdar://146075994</a>

Reviewed by Tyler Wilcock.

When we have a text marker at the root web area with offset 0, nothing can come before it.
With this knowledge, we can add a fast path to partialOrderByTraversal, so that we don&apos;t
have to walk the entire AX tree to determine ordering.

* LayoutTests/accessibility/ax-thread-text-apis/range-from-webarea-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/range-from-webarea.html: Added.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isRootWebArea const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::partialOrderByTraversal const):

Canonical link: <a href="https://commits.webkit.org/291578@main">https://commits.webkit.org/291578@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cbb04037a9c3e43fd41ace8a7757dcd42298629f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93306 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12864 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2591 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98305 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43832 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95356 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13151 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21319 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71306 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28703 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96308 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9871 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84401 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51639 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9562 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2033 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43146 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79848 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2048 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100337 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20358 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14903 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80327 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20610 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80321 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79642 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19792 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24188 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1522 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13481 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20342 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25519 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20029 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23489 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21770 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -1406,6 +1406,16 @@ std::partial_ordering AXTextMarker::partialOrderByTraversal(const AXTextMarker&
+
+    RefPtr object2 = other.object();
+    if (object2 && !other.offset() && object2->isRootWebArea())
+        return std::partial_ordering::greater;
```
To reduce ref-count churn:

1. Rather than making a `RefPtr object1 = object()` here, and then `RefPtr current = object();`, maybe we can just move `RefPtr current = object();` up to this location?
2. For `object2`, maybe we only create `RefPtr object2` when `!other.offset()`? Also, what do you think about `otherObject` as a name? That pairs nicely with the fact that the other `AXTextMarkerRange` is called `other` — `otherObject` feels intuitive based on that.

hoffmanjoshua replied:
Updated addressing both of these—thanks for the feedback!

hoffmanjoshua committed:
AX: Add fast path for text marker ordering when one marker is the start of the root web area
https://bugs.webkit.org/show_bug.cgi?id=289034
rdar://146075994

Reviewed by Tyler Wilcock.

When we have a text marker at the root web area with offset 0, nothing can come before it.
With this knowledge, we can add a fast path to partialOrderByTraversal, so that we don't
have to walk the entire AX tree to determine ordering.

* LayoutTests/accessibility/ax-thread-text-apis/range-from-webarea-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/range-from-webarea.html: Added.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isRootWebArea const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::partialOrderByTraversal const):

Canonical link: https://commits.webkit.org/291578@main
